File: CH 01 Code/Native/Core/Thermometer.scala
package Native.Core {
	
	sealed trait Temperature
	case class Kelvin(value: Float) extends Temperature
	case class Celsius(value: Float) extends Temperature
	
	object Thermometer {
		def read(): Temperature = {
			println("Thermometer returned data.")
			Kelvin(100.0f)
		}
	}
	
	object HighAccuracyThermometer {
		def read() = {
			println("HA Thermometer returned data.")
			100.0001f
		}
	}
	
	object Utils {
		def toCelsius(data: Float) : Float = data - 273.15f
		def toCelsius(data: Native.Core.Temperature) : Float =
			data match {
				case Native.Core.Kelvin(v)  => toCelsius(v)
				case Native.Core.Celsius(v) => v
			}
	}
}

File: Self-Published-Edition/BookSamples/CH01/L1_3/Native/Core/Thermometer.hs
module Native.Core.Thermometer where

data Temperature = Kelvin Float
                 | Celsius Float

r = putStrLn "Read"

read :: String -> IO Temperature
read "T-201A"  = (r >>) $ return $ Kelvin 323.5
read "T-35B"   = (r >>) $ return $ Kelvin 325.1
read "T-201AC" = (r >>) $ return $ Kelvin 643.8
read n = error $ "Thermometer not found: " ++ n

